### PR TITLE
Add docs for finch-iteratee module

### PR DESCRIPTION
**Pull request by ImLiar**
Related to https://github.com/finagle/finch/issues/822

### CODE REVIEW

File: docs/src/main/tut/user-guide.md [hunk 1]
```diff
@@ -803,6 +803,115 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+- `Enumerator` is a "lazy storage" with all the chunks that already were received or about to be received in a future
+- `Enumeratee` is a `map` part and could be used to transform chunks in `Enumerator`
+- `Iteratee` is a `reduce` part that runs computation and processes over chunks in `Enumerator`
+
+**Decoding**
+
+Currently JSON decoding is supported only in `finch-circe` module and implemented
+using `circe-streaming` library.
+
+To make it happen finch out of the box provides few instances for `io.finch.iteratee.Enumerate` abstraction:
```
Comment (vkostyukov) [inline docs/src/main/tut/user-guide.md]: How about:

Finch plugs in iteratee-powered decoding support via the `io.finch.iteratee.Enumerate` type-class.

File: docs/src/main/tut/user-guide.md [hunk 2]
```diff
@@ -803,6 +803,115 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+/**
+  * Backpressure with HTTP 1.1 could be implemented only in a way of closing a connection.
+  * If you don't want to consume more data, you could throw any exception while processing `Enumerator`,
+  * it'll close a connection.
+  *
+  * In this example connection going to be closed as soon as 10 elements has been stored.
+  */
+val backpressureJSON = post("bar" :: enumeratorJsonBody[Foo]) { (enumerator: Enumerator[Future, Foo]) =>
+   val iteratee = Iteratee.take[Future, Foo](10).flatMapM(fs: Vector[Foo] => {
+      storeFoos(fs).map(_ => throw new InterruptedException)
```
Comment (vkostyukov) [inline docs/src/main/tut/user-guide.md]: What's `storeFoos`? What do you think about just discarding `fs` (ignoring it)?

File: docs/src/main/tut/user-guide.md [hunk 3]
```diff
@@ -803,6 +803,115 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+)
+
+Http.server
+    .withStreaming(enabled = true) //don't forget to enable streaming support
+    .serve(":8080", (decodingJSON :+: backpressureJSON :+: bufEnumerator).toService)
+```
+
+**Encoding**
+
+Beside decoding of input stream, it's possible to make output stream with enumerator using
```
Comment (vkostyukov) [inline docs/src/main/tut/user-guide.md]: s/using/serving

File: docs/src/main/tut/user-guide.md (hunk 2 above)
Comment (ImLiar) [reply docs/src/main/tut/user-guide.md]: I wanted to show that processing is interrupted if underlying future fails

**Comment by ImLiar** [comment]
@vkostyukov done

File: docs/src/main/tut/user-guide.md [hunk 4]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+import io.circe.generic.auto._
+import io.finch._
+import io.finch.circe._
+import io.finch.iteratee._
+import io.iteratee._
+
+case class Foo(x: Int)
+
+val streamingEndpoint: Endpoint[Enumerator[Future, Foo]] = get("stream") {
+   Enumerator[Future, Foo].enumList(List(Foo(1), Foo(2))
```
Comment (vkostyukov) [inline docs/src/main/tut/user-guide.md]: Sorry, I missed it in the first place: should we wrap it with `Ok()`?

File: docs/src/main/tut/user-guide.md [hunk 5]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
 If no other `Encode[Exception]` is available, Finch provides a fallthrough of `Encode.Aux[Exception, ?]`
 that will return an empty content body.
 
+### Streaming
+
+`finch-iteratee` module enables high-level support of chunked request and response streaming using
```
Comment (rpless) [inline docs/src/main/tut/user-guide.md]: Can we add `The` to the start of this sentence.

File: docs/src/main/tut/user-guide.md [hunk 6]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+[iteratee.io](https://github.com/travisbrown/iteratee) and `circe-streaming` libraries. It allows to encode and decode
```
Comment (rpless) [inline docs/src/main/tut/user-guide.md]: Maybe `It allows encoding and decoding newline delimited JSON streams, but also supports binary and text streaming.`

Also s/delimeted/delimited whether we like my above suggestion or not.

File: docs/src/main/tut/user-guide.md [hunk 7]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+newline delimeted JSON streams, but also supports binary and text streaming.  
+Concept of [iteratee](https://en.wikipedia.org/wiki/Iteratee) could be complicated for understanding, but
```
Comment (rpless) [inline docs/src/main/tut/user-guide.md]: Maybe:
``` Conceptually iteratee can be difficult to understand, but proves itself a very powerful abstraction for processing sequential data.
```

File: docs/src/main/tut/user-guide.md [hunk 8]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+proves itself very powerful and useful abstraction over processing sequential data. API of `iteratee.io` library is transparent and covers most of
```
Comment (rpless) [inline docs/src/main/tut/user-guide.md]: I think we need to add a `the` to the start of `API of iteratee.io`.

File: docs/src/main/tut/user-guide.md [hunk 9]
```diff
@@ -803,6 +803,116 @@ implicit val encodeExceptionCirce: Encoder[Exception] = Encoder.instance(e =>
...
+trait Enumerate[A] {
+
+  type ContentType <: String
+
+  def apply(enumerator: Enumerator[Future, Buf], cs: Charset): Enumerator[Future, A]
+}
+
+```
+
+Here you could see an example how to work with decoding enumerators:
```
Comment (rpless) [inline docs/src/main/tut/user-guide.md]: I think this should be `can` instead of `could`.

**Comment by ImLiar** [comment]
@rpless done

**Comment by vkostyukov** [comment]
@ImLiar Do you want to rebase or you're fine with squash-and-merge?

**Comment by ImLiar** [comment]
@vkostyukov feel free to squash.
amazing GitHub feature, saves some time and money: both for contributor and travis 😄

**Pull request merged**